clang/test/CIR/Transforms/ternary-fold.cpp

```diff
@@ -0,0 +1,56 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir -mmlir --mlir-print-ir-before=cir-simplify %s -o %t1.cir 2>&1 | FileCheck -check-prefix=CIR-BEFORE %s
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir -mmlir --mlir-print-ir-after=cir-simplify %s -o %t2.cir 2>&1 | FileCheck -check-prefix=CIR-AFTER %s
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-llvm %s -o %t.ll 
+// RUN: FileCheck --input-file=%t.ll --check-prefix=LLVM %s
+
+int test1(bool x) {
+  return x ? 1 : 2;
+}
+
+//      CIR-BEFORE: cir.func @_Z5test1b
+//      CIR-BEFORE:   %{{.+}} = cir.ternary(%{{.+}}, true {
+// CIR-BEFORE-NEXT:     %[[#A:]] = cir.const #cir.int<1> : !s32i
+// CIR-BEFORE-NEXT:     cir.yield %[[#A]] : !s32i
+// CIR-BEFORE-NEXT:   }, false {
+// CIR-BEFORE-NEXT:     %[[#B:]] = cir.const #cir.int<2> : !s32i
+// CIR-BEFORE-NEXT:     cir.yield %[[#B]] : !s32i
+// CIR-BEFORE-NEXT:   }) : (!cir.bool) -> !s32i
+//      CIR-BEFORE: }
+
+//      CIR-AFTER: cir.func @_Z5test1b
+//      CIR-AFTER:   %[[#A:]] = cir.const #cir.int<1> : !s32i
+// CIR-AFTER-NEXT:   %[[#B:]] = cir.const #cir.int<2> : !s32i
+// CIR-AFTER-NEXT:   %{{.+}} = cir.select if %{{.+}} then %[[#A]] else %[[#B]] : (!cir.bool, !s32i, !s32i) -> !s32i
+//      CIR-AFTER: }
+
+// LLVM: define dso_local i32 @_Z5test1b
+// LLVM:   %{{.+}} = select i1 %{{.+}}, i32 1, i32 2
+// LLVM: }
+
+int test2(bool x, int a, int b) {
+  return x ? a : b;
+}
+
+//      CIR-BEFORE: cir.func @_Z5test2bii
+//      CIR-BEFORE:   %{{.+}} = cir.ternary(%{{.+}}, true {
+// CIR-BEFORE-NEXT:     %[[#A:]] = cir.load %{{.+}} : !cir.ptr<!s32i>, !s32i
+// CIR-BEFORE-NEXT:     cir.yield %[[#A]] : !s32i
+// CIR-BEFORE-NEXT:   }, false {
+// CIR-BEFORE-NEXT:     %[[#B:]] = cir.load %2 : !cir.ptr<!s32i>, !s32i
+// CIR-BEFORE-NEXT:     cir.yield %[[#B]] : !s32i
+// CIR-BEFORE-NEXT:   }) : (!cir.bool) -> !s32i
+//      CIR-BEFORE: }
+
+
+//      CIR-AFTER: cir.func @_Z5test2bii
+//      CIR-AFTER:   %[[#A:]] = cir.load %{{.+}} : !cir.ptr<!s32i>, !s32i
+// CIR-AFTER-NEXT:   %[[#B:]] = cir.load %{{.+}} : !cir.ptr<!s32i>, !s32i
```

Comment:
By hoisting the loads we are evaluating a load that wouldn't be evaluated otherwise, better not touch loads in general as part of this PR - this is something we want mem2reg to do the right job first (and/or a combo of other optimizations)

Reply:
The hoisting rules are hard-coded in this PR to make simple C/C++ ternary expressions eventually lower to a select operation. By "simple" I refer to those ternary expressions that yields literal constants or id-expressions on both of its branches.

At first I thought this is unnecessary since `mem2reg` should do the job eliminating the loads. But then I realize that `mem2reg` only works on flat control flow and thus it cannot work before `cir.ternary` is flattened. Do we have a workaround for this?

Reply:
My point is more general, it's not safe to do the hoisting, you are forcing execution of something that should only execute conditionally - it's also too early in the pipeline to speculate without data evidence of sorts (heuristics, PGO, etc). It's also possible that LLVM optimizations will already simplify this further for us, with their tons of existing passes, have you look at the final LLVM IR for the case you care after passing -O1 and above?

> But then I realize that mem2reg only works on flat control flow and thus it cannot work before cir.ternary is flattened. Do we have a workaround for this?

Unfortunately it's a known limitation! We can still make some assumptions about some loads/stores when it makes sense, I just don't think we need it right here.

Reply:
My suggestions: let's make this PR a bit more simple and not handle loads right now.

Reply:
Updated.